### ISSUE TITLE
Ranking issue

### ISSUE THREAD

**Issue by acasar** [teamtnt/tntsearch#40]
I'm having some ranking issues. When I use multiple words in the search string, the result containing all of the terms is not at the top. When using only one word the results are ok. How does searching with multiple terms work?

If I copy-paste the title of the article in the search box I would expect it to be the top result but it seems that it's always outranked by all articles containing only one of the words.

**Comment by acasar** [comment]
I'm using the normal ``search`` method and no tokenizer.

**Comment by nticaric** [comment]
Some of the words outweigh others because they occur more times in a document. A common case are the words `the`, `and`, `a`, so you might treat them as stopwords.

**Comment by nticaric** [comment]
In case you need `all` of the words to occur, you can use `booleanSearch`. To better undesrtand the ranking algoritham for the `search` method take a look here https://en.wikipedia.org/wiki/Okapi_BM25

The algorithm is slightly modified because we don't take the  document length into consideration

**Comment by acasar** [comment]
Don't necessarily want all of the words to appear, but currently one word match outranks everything. This gif explains it (using some dummy data):

![bug](https://cloud.githubusercontent.com/assets/6329543/18650759/165be75a-7ec7-11e6-8443-04317dd65f12.gif)

**Comment by nticaric** [comment]
Try to treat `et` and `set` as stopwords. I'm not sure what database you use, but if you use a regular `where in` clause it might mix up the order

**Comment by nticaric** [comment]
Just in case you forgot, something like this is necessary for correct ranking
```
$ids = implode(",", $res['ids']);
$articles = Article::whereIn('id', $res['ids'])->orderByRaw("FIELD (ID, $ids)")->get();
```

**Comment by acasar** [comment]
I've got the correct orderBy and I added the stopwords but still no improvement. It's really weird - seems almost impossible for results to be so wrong.

I'll do some more digging tomorrow and I'll report here if I find a solution. Thanks!

**Comment by nticaric** [comment]
Glad to help. Can you share your dummy data so I can try to recreate the problem and see whats going on

**Comment by acasar** [comment]
Sure! Attached is the SQL to build up the ``sp_catalogue_products`` table. I am indexing "title" and "abstract" fields. I also tried to index title only, but didn't help with the results much.

[sp_catalogue_products.txt](https://github.com/teamtnt/tntsearch/files/481313/sp_catalogue_products.txt)

As I said, I'll try to solve this issue tomorrow, but any help would be very appreciated. :)

**Comment by nticaric** [comment]
Just tried it out with your dataset, and the documents are ordered correctly for me. 
In order to speed up the setup, I used laravel scout with tntsearch driver

**Comment by acasar** [comment]
@nticaric I think I found the issue. I started with an empty database and therefore I just created an empty index using ``$tnt->createIndex('products.index');`` When I started adding new items, I just called ``$index->insert($data);`` for each new row in the database.

That created the documents in the index, but the key ``total_documents`` in the ``info`` table was never created. Therefore the ranking is off, since ``total_documents`` is always null.

I think that the ``total_documents`` key should be inserted in the ``createIndex`` method rather than in the ``run`` method? I think that if you're starting with an empty database you shouldn't be forced to call the ``run`` method?
